### PR TITLE
[don't merge yet] Scalar training: Spiral dataset optimizations

### PR DESCRIPTION
**Pull request by r-chong**
Significant amount of changes:
- Optimize spiral dataset training
    - BCEWithLogitsLoss — symmetric gradients, no log(sigmoid) numerical path
    - He initialization — proper weight scale for ReLU-family activations
    - Zero bias init — prevents dead neurons at startup
    - LeakyReLU — ensures gradients always flow, neurons can revive
    - Adam optimizer — adaptive learning rates + momentum smooth out the noisy loss landscape
    - Multi-layer architecture — depth to represent the spiral boundary

### CODE REVIEW

File: packages/tstorch/src/scalar_functions.ts [hunk 1]
```diff
@@ -111,6 +111,20 @@ export class Relu extends ScalarFunction {
     }
 }
 
+const LEAKY_SLOPE = 0.01;
+
+export class LeakyRelu extends ScalarFunction {
+    static forward(ctx: Context, a: number): number {
+        ctx.saveForBackward(a);
+        return a > 0 ? a : LEAKY_SLOPE * a;
```
Comment (MankyDanky) [inline packages/tstorch/src/scalar_functions.ts:119]: Will using a constant outside the scope of the Class LeakyRelu mess with the pure function requirement of 3.1? I dont think it will but can we test this before merging to be safe?

File: packages/tstorch/src/optimizer.ts [hunk 2]
```diff
@@ -10,45 +10,91 @@ export class Optimizer {
...
-        if (p.value instanceof Scalar) {
+    step() {
+        this.t++;
+        for (const p of this.parameters) {
+            if (!(p.value instanceof Scalar)) continue;
             const grad = p.value.derivative ?? 0;
-            p.value.data -= this.lr * grad;
+
+            let mi = this.m.get(p)!;
+            let vi = this.v.get(p)!;
```
Comment (MankyDanky) [inline packages/tstorch/src/optimizer.ts:86]: Is there any risk of this being uninitialized and given an error? If so should we do lazy initialization with `?? 0`? If there is no risk then ignore this

File: packages/tstorch/src/scalar_functions.ts (hunk 1 above)
Comment (r-chong) [reply packages/tstorch/src/scalar_functions.ts:119]: Yes

File: packages/tstorch/src/optimizer.ts (hunk 2 above)
Comment (r-chong) [reply packages/tstorch/src/optimizer.ts:86]: Will test
